File: app/src/main/java/com/mobileclaw/app/orchestration/OrchestrationTypes.kt
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mobileclaw.app.orchestration

/** A single step in an execution plan. */
data class PlanStep(
  val id: String,
  val description: String,
  val skillName: String? = null,
  val toolName: String? = null,
  val toolArgs: Map<String, String> = emptyMap(),
  val dependsOn: List<String> = emptyList(),
)

/** A structured execution plan produced by the Planner. */
data class ExecutionPlan(
  val goal: String,
  val reasoning: String,
  val steps: List<PlanStep>,
  /**
   * Concrete outcomes the user would call "success" — used by the evaluator as a rubric.
   * Empty means the evaluator will fall back to the old holistic check.
   */
  val successCriteria: List<String> = emptyList(),
)

/** Status of a single plan step during execution. */
enum class StepStatus {
  PENDING,
  RUNNING,
  COMPLETED,
  FAILED,
  SKIPPED,
}

/** Result of executing a single plan step. */
data class StepResult(
  val stepId: String,
  val status: StepStatus,
  val output: String = "",
  val error: String? = null,
  val durationMs: Long = 0,
)

/** Overall status of the orchestration loop. */
enum class OrchestrationStatus {
  IDLE,
  PLANNING,
  EXECUTING,
  EVALUATING,
  REPAIRING,
  REPLANNING,
  FORMATTING,
  COMPLETED,
  CANCELLED,
  ERROR,
}

/** Full state of an orchestration run, exposed via StateFlow. */
data class OrchestrationState(
  val status: OrchestrationStatus = OrchestrationStatus.IDLE,
  val plan: ExecutionPlan? = null,
  val stepResults: Map<String, StepResult> = emptyMap(),
  val evaluation: EvaluationResult? = null,
  val iteration: Int = 0,
  val maxIterations: Int = 3,
  val finalOutput: String? = null,
  val finalOutputIsHtml: Boolean = false,
  val error: String? = null,
  val memoryRecalled: Boolean? = null,
  /** Whether thinking/CoT was enabled for each LLM phase during this run. */
  val thinkingByPhase: Map<String, Boolean> = emptyMap(),
)

/** Result of the self-evaluation module. */
data class EvaluationResult(
  val goalAchieved: Boolean,
  val assessment: String,
  val missingItems: List<String> = emptyList(),
  val shouldReplan: Boolean = false,
  /** Criteria the rubric check marked as failed (when rubric was used). */
  val failedCriteria: List<String> = emptyList(),
)

/** Lightweight skill descriptor for planning prompts. */
data class SkillSummary(
  val name: String,
  val description: String,
  val instructions: String = "",
  /** "base" = always in planner catalog. "deferred" = name-only until search-skills loads it. */
  val tier: String = "base",
)

/** Result of diagnosing a failed skill step. */
data class DiagnosticResult(
  val diagnosis: String,
  val fixType: String, // "retry_with_different_args", "use_alternative_skill", "skip", "unfixable"
  val alternativeSkillName: String? = null,
  val alternativeArgs: Map<String, String> = emptyMap(),
  val updatedInstructions: String? = null,
)

File: ios-app-wip/MobileClaw/Orchestration/OrchestrationTypes.swift
import Foundation

// 1:1 port of android-app/.../orchestration/OrchestrationTypes.kt

public struct PlanStep: Sendable, Equatable {
    public let id: String
    public let description: String
    public let skillName: String?
    public let toolName: String?
    public let toolArgs: [String: String]
    public let dependsOn: [String]

    public init(id: String, description: String, skillName: String? = nil,
                toolName: String? = nil, toolArgs: [String: String] = [:],
                dependsOn: [String] = []) {
        self.id = id; self.description = description; self.skillName = skillName
        self.toolName = toolName; self.toolArgs = toolArgs; self.dependsOn = dependsOn
    }
}

public struct ExecutionPlan: Sendable, Equatable {
    public let goal: String
    public let reasoning: String
    public let steps: [PlanStep]
    public let successCriteria: [String]
    public init(goal: String, reasoning: String, steps: [PlanStep], successCriteria: [String] = []) {
        self.goal = goal; self.reasoning = reasoning; self.steps = steps; self.successCriteria = successCriteria
    }
}

public enum StepStatus: String, Sendable {
    case pending, running, completed, failed, skipped
}

public struct StepResult: Sendable, Equatable {
    public let stepId: String
    public let status: StepStatus
    public let output: String
    public let error: String?
    public let durationMs: Int64
    public init(stepId: String, status: StepStatus, output: String = "", error: String? = nil, durationMs: Int64 = 0) {
        self.stepId = stepId; self.status = status; self.output = output; self.error = error; self.durationMs = durationMs
    }
}

public enum OrchestrationStatus: String, Sendable {
    case idle, planning, executing, evaluating, repairing, replanning, formatting, completed, cancelled, error
}

public struct OrchestrationState: Sendable {
    public var status: OrchestrationStatus = .idle
    public var plan: ExecutionPlan?
    public var stepResults: [String: StepResult] = [:]
    public var evaluation: EvaluationResult?
    public var iteration: Int = 0
    public var maxIterations: Int = 3
    public var finalOutput: String?
    public var finalOutputIsHtml: Bool = false
    public var error: String?
    public var memoryRecalled: Bool?
    public var thinkingByPhase: [String: Bool] = [:]
    public init() {}
}

public struct EvaluationResult: Sendable, Equatable {
    public let goalAchieved: Bool
    public let assessment: String
    public let missingItems: [String]
    public let shouldReplan: Bool
    public let failedCriteria: [String]
    public init(goalAchieved: Bool, assessment: String, missingItems: [String] = [],
                shouldReplan: Bool = false, failedCriteria: [String] = []) {
        self.goalAchieved = goalAchieved; self.assessment = assessment
        self.missingItems = missingItems; self.shouldReplan = shouldReplan
        self.failedCriteria = failedCriteria
    }
}

public struct SkillSummary: Sendable, Equatable {
    public let name: String
    public let description: String
    public let instructions: String
    /// "base" = always in planner catalog. "deferred" = name-only until search-skills loads it.
    public let tier: String
    public init(name: String, description: String, instructions: String = "", tier: String = "base") {
        self.name = name; self.description = description; self.instructions = instructions; self.tier = tier
    }
}

public struct DiagnosticResult: Sendable, Equatable {
    public let diagnosis: String
    public let fixType: String   // retry_with_different_args / use_alternative_skill / skip / unfixable
    public let alternativeSkillName: String?
    public let alternativeArgs: [String: String]
    public let updatedInstructions: String?
    public init(diagnosis: String, fixType: String, alternativeSkillName: String? = nil,
                alternativeArgs: [String: String] = [:], updatedInstructions: String? = nil) {
        self.diagnosis = diagnosis; self.fixType = fixType
        self.alternativeSkillName = alternativeSkillName; self.alternativeArgs = alternativeArgs
        self.updatedInstructions = updatedInstructions
    }
}
